Sources/Prelude/Refinements.swift

```diff
@@ -0,0 +1,207 @@
+//
+// This source file is part of Prelude, an open source project by Wayfair
+//
+// Copyright (c) 2018 Wayfair, LLC.
+// Licensed under the 2-Clause BSD License
+//
+// See LICENSE.md for license information
+//
+
+/// protocol that represents a rule that can be used to “refine” values of type `RefinedType`. Implement the `static Refinement.isValid(_:)` function to describe which values are allowed.
+public protocol Refinement {
+    /// the type being “refined”
+    associatedtype RefinedType
```

Comment:
I’ve been thinking about it and I’m wondering if it would be more clear to rename this to `BaseType`. It’d be one less type in this library that starts with `Refine…` and that might help people pick it up. 

There’s the OO connotation of a “base class” which is maybe a little problematic, but I’m thinking that might not be a huge deal.